File: laocai-bot-spring-boot-starter/src/main/java/indi/kyson/laocai/bot/model/segment/OutgoingReplySegment.java
package indi.kyson.laocai.bot.model.segment;

import com.fasterxml.jackson.annotation.JsonIgnore;

import java.util.Objects;

/**
 * 出站的引用回复消息段。
 * <p>
 * 发送时只需要指明被回复消息的 messageSeq，协议字段名为 shouldReply。
 */
public final class OutgoingReplySegment implements Segment {

    record Data(Long messageSeq) {
        OutgoingReplySegment toSegment() {
            return new OutgoingReplySegment(this);
        }
    }

    private final Data data;

    private OutgoingReplySegment(Data data) {
        this.data = data;
    }

    public static OutgoingReplySegment of(Long messageSeq) {
        return new OutgoingReplySegment(new Data(Objects.requireNonNull(messageSeq)));
    }

    @Override
    public String getType() {
        return "reply";
    }

    @Override
    public Data getData() {
        return data;
    }

    @JsonIgnore
    public Long getMessageSeq() {
        return data.messageSeq();
    }
}

File: laocai-bot-core/src/main/kotlin/indi/kyson/laocai/bot/core/segment/OutgoingReplySegment.kt
package indi.kyson.laocai.bot.core.segment

import com.fasterxml.jackson.annotation.JsonIgnore

/**
 * 出站的引用回复消息段。
 *
 * 发送时只需要指明被回复消息的 messageSeq，协议字段名为 shouldReply。
 */
class OutgoingReplySegment private constructor(override val data: Data) : Segment {

    data class Data(val messageSeq: Long) {
        fun toSegment(): OutgoingReplySegment = OutgoingReplySegment(this)
    }

    override val type: String
        get() = "reply"

    @get:JsonIgnore
    val messageSeq: Long
        get() = data.messageSeq

    companion object {
        @JvmStatic
        fun of(messageSeq: Long): OutgoingReplySegment = OutgoingReplySegment(Data(messageSeq))
    }
}
